arangod/Aql/Function.h

```diff
@@ -50,36 +54,62 @@ struct Function {
     /// cache
     Cacheable = 2,
 
-    /// @brief whether or not the function may be executed on DB servers
-    CanRunOnDBServer = 4,
+    // the following two flags control separately if a function can be pushed
+    // down to DB servers for execution. In almost cases we want to have the
+    // "Cluster" flag set for functions that have the "OneShard" flag set,
+    // but this is currently only enforced via assertions - and we may want
+    // to change this in the future so we can have functions that are _not_
+    // pushed down to DB servers in OneShard mode but are in normal Cluster
+    // mode.
+
+    /// @brief whether or not the function may be executed on DB servers, 
+    /// general cluster case (non-OneShard)
+    /// note: in almost all circumstances it is also useful to set the flag
+    /// CanRunOnDBServerOneShard in addition!
+    CanRunOnDBServerCluster = 4,
+    
+    /// @brief whether or not the function may be executed on DB servers,
+    /// OneShard databases. 
+    /// note: this flag must be set in addition to CanRunOnDBServerCluster
+    /// to make a function run on DB servers in OneShard mode! 
+    CanRunOnDBServerOneShard = 8,
+    
+    /// @brief whether or not the function may read documents from the database
+    CanReadDocuments = 16,
 
     /// @brief exclude the function from being evaluated during AST
     /// optimizations evaluation of function will only happen at query runtime
-    NoEval = 8,
+    NoEval = 32,
   };
 
   /// @brief helper for building flags
   template <typename... Args>
-  static inline std::underlying_type<Flags>::type makeFlags(Flags flag, Args... args) {
+  static inline std::underlying_type<Flags>::type makeFlags(Flags flag, Args... args) noexcept {
     return static_cast<std::underlying_type<Flags>::type>(flag) + makeFlags(args...);
```

Comment:
```suggestion
    return static_cast<std::underlying_type<Flags>::type>(flag) | makeFlags(args...);
```
ATM adding the same flag twice has unintended consequences.